### ISSUE TITLE
Case Sensitivity - show warning?

### ISSUE THREAD

**Issue by matanshukry** [gotwarlost/istanbul#368]
It seems istanbul is using the require() global function for code coverage.
As a mistake, in one of my models I used a capital letter, as follows: 

```
require('my_module');
```
And in one case I used:
```
require('my_Module');
```

As a result, some functions were reported as not covered by istanbul (I verified they were running by printing a console message in them).

It would be nice if istanbul could check the case sensitivity of the path against the file loaded, and possibly warn about it. 

p.s.
 * I'm currently using windows, which is why the require even worked, of course.
 * There is an issue on node regarding require and case sensitivity (https://github.com/joyent/node/issues/2621 ), but in my opinion the 'require' function shouldn't be changed.

**Comment by davglass** [comment]
I don't think this is actually `istanbul`'s job to do this warning, we would have to do a ton of logic to determine if a module path was in a different case (it could be any part of the path that could be different). We would have to process every file in a really hacky way in order to even guess at what you wanted us to warn about. I think tossing the coverage in, like we do, with the "unknown" file is the best way as it's not really hacky.

**Comment by matanshukry** [comment]
Not really sure what you mean by "tossing the coverage in".
Currently, the result is nothing is logging a warning, and when looking at the coverage reported by istanbul, some code shows as not covered, even though it is.

This makes the user believe the problem is with istanbul, and not with the case sensitive issue (at least until he finds out)

Also, I don't think the logic should be very complex. A simple map of lowrecased filenames with an existence test should suffice. Something like:

```
var uniqueFiles = {};
..
Module._extensions[ext] = function (module, filename) {
  var uniqueFilename = path.resolve(fileName).toLowerCase();
  if (uniqueFiles[uniqueFilename]) {
    // log warning; We have two files with the same unique name
    // 'uniqueFilename' and 'uniqueFiles[uniqueFilename]'
  } else {
    uniqueFiles[uniqueFilename] = true;
  }
  ...
};
```

Note: I used a map and not an array in here since as far as I know, the object lookup performance should be better than array. It can be changed to array though, of course.

**Comment by rcdmk** [comment]
I had this problem also and it took me many hours to figure out that a file was being "required" with one letter in a different case than the real file name.

I think a warning is not realy necessary. What I think is the best solution is to ignore the file case on Windows enviroments.

**Comment by gotwarlost** [comment]
Changed the subject so I can find this issue later :)

When you have 2 require statements with different casing, node actually loads the module twice with 2 different names into its cache. (IMO, this is an issue if there is module-level state that the library relies on).

All of istanbul's require hooks and coverage reporting should actually work correctly.

The reason it doesn't work is bizzare. istanbul's file matcher is implemented as a cache of all seen file names and for istanbul to instrument a file the file needs to be part of the filename cache that it loads and in this case, it isn't.

The implementation of istanbul's file matcher has always been suspect from day 1 and I never bothered to fix it. It is slow and bloated because it looks at all filesystem files whether they are used or not. This needs to be changed to nuke the internal cache and actually do a pattern match when a filename is encountered.

Working on this soon since it will fix other performance problems as well.

**Comment by sjodatti** [comment]
I am facing same issue. Is there any update on this fix?
